src/module.cpp

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright 2019 IBM Corp. and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sstream>
+
+#include "module.hpp"
+
+namespace winter {
+
+Module::Module(const AbstractModule& abstract, Environment& env)
+    : _imports(abstract.imports()), _exports(abstract.exports()), _memories(abstract.memories()), _env(&env) {
+    _import_func_sigs.reserve(abstract.funcs().size());
+    _funcs.reserve(abstract.funcs().size());
+    for (const AbstractFunc& func : abstract.funcs()) {
+        if (!func.is_import) {
+            _import_func_sigs.emplace_back(nullptr);
+            _funcs.emplace_back(UnlinkedFunc::instantiate(func, env));
+        } else {
+            _import_func_sigs.emplace_back(&env.types().sig(func.sig));
+            _funcs.emplace_back(nullptr);
+        }
+    }
+
+    for (AbstractMemory& mem : _memories) {
+        if (mem.is_shared && !mem.is_import) {
+            _shared_memories.emplace_back(new Memory(mem));
+        } else {
+            _shared_memories.emplace_back(nullptr);
+        }
+    }
+}
+
+LinkedFunc* ImportMultiModule::find_func(const Import& import) const {
+    for (ImportModule* module : _modules) {
+        LinkedFunc* func = module->find_func(import);
+        if (func)
+            return func;
+    }
```

Comment:
The problem is that the thing we're finding is the _result_ of the `module->find_func(import)` invocation, not `import` itself. Additionally, calling that function multiple times is probably not a great idea, since it's potentially user-implemented.

Reply:
I see.